File: Kernel Simulator/Network/Mail/MailTransfer/MailTransfer.vb
﻿
'    Kernel Simulator  Copyright (C) 2018-2020  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.IO
Imports System.Text
Imports MailKit
Imports MimeKit
Imports MimeKit.Cryptography
Imports MimeKit.Text

Public Module MailTransfer

    ''' <summary>
    ''' Prints content of message to console
    ''' </summary>
    ''' <param name="MessageNum">Message number</param>
    Public Sub MailPrintMessage(ByVal MessageNum As Integer, Optional ByVal Decrypt As Boolean = False)
        Dim Message As Integer = MessageNum - 1
        Dim MaxMessagesIndex As Integer = IMAP_Messages.Count - 1
        Wdbg("I", "Message number {0}", Message)
        If Message < 0 Then
            Wdbg("E", "Trying to access message 0 or less than 0.")
            W(DoTranslation("Message number may not be negative or zero.", currentLang), True, ColTypes.Err)
            Exit Sub
        ElseIf Message > MaxMessagesIndex Then
            Wdbg("E", "Message {0} not in list. It was larger than MaxMessagesIndex ({1})", Message, MaxMessagesIndex)
            W(DoTranslation("Message specified is not found.", currentLang), True, ColTypes.Err)
            Exit Sub
        End If

        SyncLock IMAP_Client.SyncRoot
            'Get message
            Wdbg("I", "Getting message...")
            Dim Msg As MimeMessage
            If Not IMAP_CurrentDirectory = "" And Not IMAP_CurrentDirectory = "Inbox" Then
                Dim Dir As MailFolder = OpenFolder(IMAP_CurrentDirectory)
                Msg = Dir.GetMessage(IMAP_Messages(Message))
            Else
                Msg = IMAP_Client.Inbox.GetMessage(IMAP_Messages(Message))
            End If

            'Prepare view
            Console.WriteLine()

            'Print all the addresses that sent the mail
            Wdbg("I", "{0} senders.", Msg.From.Count)
            For Each Address As InternetAddress In Msg.From
                Wdbg("I", "Address: {0} ({1})", Address.Name, Address.Encoding.EncodingName)
                W(DoTranslation("- From {0}", currentLang), True, ColTypes.HelpCmd, Address.ToString)
            Next

            'Print all the addresses that received the mail
            Wdbg("I", "{0} receivers.", Msg.To.Count)
            For Each Address As InternetAddress In Msg.To
                Wdbg("I", "Address: {0} ({1})", Address.Name, Address.Encoding.EncodingName)
                W(DoTranslation("- To {0}", currentLang), True, ColTypes.HelpCmd, Address.ToString)
            Next

            'Print the date and time when the user received the mail
            Wdbg("I", "Rendering time and date of {0}.", Msg.Date.DateTime.ToString)
            W(DoTranslation("- Sent at {0} in {1}", currentLang), True, ColTypes.HelpCmd, RenderTime(Msg.Date.DateTime), RenderDate(Msg.Date.DateTime))

            'Prepare subject
            Console.WriteLine()
            Wdbg("I", "Subject length: {0}, {1}", Msg.Subject.Length, Msg.Subject)
            W($"- {Msg.Subject}", False, ColTypes.HelpCmd)

            'Write a sign after the subject if attachments are found
            Wdbg("I", "Attachments count: {0}", Msg.Attachments.Count)
            If Msg.Attachments.Count > 0 Then
                W(" - [*]", True, ColTypes.HelpCmd)
            Else
                Console.WriteLine()
            End If

            'Prepare body
            Console.WriteLine()
            Wdbg("I", "Displaying body...")
            Dim DecryptedMessage As Dictionary(Of String, MimeEntity)
            Wdbg("I", "To decrypt: {0}", Decrypt)
            If Decrypt Then
                DecryptedMessage = DecryptMessage(Msg)
                Wdbg("I", "Decrypted messages length: {0}", DecryptedMessage.Count)
                Dim DecryptedEntity As MimeEntity = DecryptedMessage("Body")
                Dim DecryptedStream As New MemoryStream
                Wdbg("I", $"Decrypted message type: {If(TypeOf DecryptedEntity Is Multipart, "Multipart", "Singlepart")}")
                If TypeOf DecryptedEntity Is Multipart Then
                    Dim MultiEntity As Multipart = CType(DecryptedEntity, Multipart)
                    Wdbg("I", $"Decrypted message entity is {If(Not IsNothing(MultiEntity), "multipart", "nothing")}")
                    If Not IsNothing(MultiEntity) Then
                        For EntityNumber As Integer = 0 To MultiEntity.Count - 1
                            Wdbg("I", $"Entity number {EntityNumber} is {If(MultiEntity(EntityNumber).IsAttachment, "an attachment", "not an attachment")}")
                            If Not MultiEntity(EntityNumber).IsAttachment Then
                                MultiEntity(EntityNumber).WriteTo(DecryptedStream, True)
                                Wdbg("I", "Written {0} bytes to stream.", DecryptedStream.Length)
                                DecryptedStream.Position = 0
                                Dim DecryptedByte(DecryptedStream.Length) As Byte
                                DecryptedStream.Read(DecryptedByte, 0, DecryptedStream.Length)
                                Wdbg("I", "Written {0} bytes to buffer.", DecryptedByte.Length)
                                W(Encoding.Default.GetString(DecryptedByte), True, ColTypes.HelpDef)
                            End If
                        Next
                    End If
                Else
                    DecryptedEntity.WriteTo(DecryptedStream, True)
                    Wdbg("I", "Written {0} bytes to stream.", DecryptedStream.Length)
                    DecryptedStream.Position = 0
                    Dim DecryptedByte(DecryptedStream.Length) As Byte
                    DecryptedStream.Read(DecryptedByte, 0, DecryptedStream.Length)
                    Wdbg("I", "Written {0} bytes to buffer.", DecryptedByte.Length)
                    W(Encoding.Default.GetString(DecryptedByte), True, ColTypes.HelpDef)
                End If
            Else
                W(Msg.GetTextBody(TextFormat.Plain), True, ColTypes.HelpDef)
            End If
            Console.WriteLine()

            'Populate attachments
#Disable Warning BC42104
            If Msg.Attachments.Count > 0 Then
                W(DoTranslation("Attachments:", currentLang), True, ColTypes.Neutral)
                Dim AttachmentEntities As New List(Of MimeEntity)
                If Decrypt Then
                    Wdbg("I", "Parsing attachments...")
                    For DecryptedEntityNumber As Integer = 0 To DecryptedMessage.Count - 1
                        Wdbg("I", "Is entity number {0} an attachment? {1}", DecryptedEntityNumber, DecryptedMessage.Keys(DecryptedEntityNumber).Contains("Attachment"))
                        Wdbg("I", "Is entity number {0} a body that is a multipart? {1}", DecryptedEntityNumber, DecryptedMessage.Keys(DecryptedEntityNumber) = "Body" And TypeOf DecryptedMessage("Body") Is Multipart)
                        If DecryptedMessage.Keys(DecryptedEntityNumber).Contains("Attachment") Then
                            Wdbg("I", "Adding entity {0} to attachment entities...", DecryptedEntityNumber)
                            AttachmentEntities.Add(DecryptedMessage.Values(DecryptedEntityNumber))
                        ElseIf DecryptedMessage.Keys(DecryptedEntityNumber) = "Body" And TypeOf DecryptedMessage("Body") Is Multipart Then
                            Dim MultiEntity As Multipart = CType(DecryptedMessage("Body"), Multipart)
                            Wdbg("I", $"Decrypted message entity is {If(Not IsNothing(MultiEntity), "multipart", "nothing")}")
                            If Not IsNothing(MultiEntity) Then
                                Wdbg("I", "{0} entities found.", MultiEntity.Count)
                                For EntityNumber As Integer = 0 To MultiEntity.Count - 1
                                    Wdbg("I", $"Entity number {EntityNumber} is {If(MultiEntity(EntityNumber).IsAttachment, "an attachment", "not an attachment")}")
                                    If MultiEntity(EntityNumber).IsAttachment Then
                                        Wdbg("I", "Adding entity {0} to attachment list...", EntityNumber)
                                        AttachmentEntities.Add(MultiEntity(EntityNumber))
                                    End If
                                Next
                            End If
                        End If
                    Next
                Else
                    AttachmentEntities = Msg.Attachments
                End If
                For Each Attachment As MimeEntity In AttachmentEntities
                    Wdbg("I", "Attachment ID: {0}", Attachment.ContentId)
                    If TypeOf Attachment Is MessagePart Then
                        Wdbg("I", "Attachment is a message.")
                        W($"- {Attachment.ContentDisposition?.FileName}", True, ColTypes.Neutral)
                    Else
                        Wdbg("I", "Attachment is a file.")
                        Dim AttachmentPart As MimePart = Attachment
                        W($"- {AttachmentPart.FileName}", True, ColTypes.Neutral)
                    End If
                Next
            End If
#Enable Warning BC42104
        End SyncLock
    End Sub

    ''' <summary>
    ''' Decrypts a message
    ''' </summary>
    ''' <param name="Text">Text part</param>
    ''' <returns>A decrypted message, or null if unsuccessful.</returns>
    Public Function DecryptMessage(ByVal Text As MimeMessage) As Dictionary(Of String, MimeEntity)
        Dim EncryptedDict As New Dictionary(Of String, MimeEntity)
        Wdbg("I", $"Encrypted message type: {If(TypeOf Text.Body Is MultipartEncrypted, "Multipart", "Singlepart")}")
        If TypeOf Text.Body Is MultipartEncrypted Then
            Dim Encrypted = CType(Text.Body, MultipartEncrypted)
            Wdbg("I", $"Message type: {If(Not IsNothing(Encrypted), "MultipartEncrypted", "Nothing")}")
            Wdbg("I", "Decrypting...")
            EncryptedDict.Add("Body", Encrypted.Decrypt(New PGPContext))
        Else
            Wdbg("W", "Trying to decrypt plain text. Returning body...")
            EncryptedDict.Add("Body", Text.Body)
        End If
        Dim AttachmentNumber As Integer = 1
        For Each TextAttachment As MimeEntity In Text.Attachments
            Wdbg("I", "Attachment number {0}", AttachmentNumber)
            Wdbg("I", $"Encrypted attachment type: {If(TypeOf TextAttachment Is MultipartEncrypted, "Multipart", "Singlepart")}")
            If TypeOf TextAttachment Is MultipartEncrypted Then
                Dim Encrypted = CType(TextAttachment, MultipartEncrypted)
                Wdbg("I", $"Attachment type: {If(Not IsNothing(Encrypted), "MultipartEncrypted", "Nothing")}")
                Wdbg("I", "Decrypting...")
                EncryptedDict.Add("Attachment " & AttachmentNumber, Encrypted.Decrypt(New PGPContext))
            Else
                Wdbg("W", "Trying to decrypt plain attachment. Returning body...")
                EncryptedDict.Add("Attachment " & AttachmentNumber, TextAttachment)
            End If
            AttachmentNumber += 1
        Next
        Return EncryptedDict
    End Function

    ''' <summary>
    ''' Sends a message
    ''' </summary>
    ''' <param name="Recipient">Recipient name</param>
    ''' <param name="Subject">Subject</param>
    ''' <param name="Body">Body (only text. See <see cref="MailSendMessage(String, String, MimeEntity)"/> for more.)</param>
    ''' <returns>True if successful; False if unsuccessful.</returns>
    Public Function MailSendMessage(ByVal Recipient As String, ByVal Subject As String, ByVal Body As String) As String
        SyncLock SMTP_Client.SyncRoot
            Try
                Dim FinalMessage As New MimeMessage
                FinalMessage.From.Add(MailboxAddress.Parse(Mail_Authentication.UserName))
                Wdbg("I", "Added sender to FinalMessage.From.")
                FinalMessage.To.Add(MailboxAddress.Parse(Recipient))
                Wdbg("I", "Added address to FinalMessage.To.")
                FinalMessage.Subject = Subject
                Wdbg("I", "Added subject to FinalMessage.Subject.")
                FinalMessage.Body = New TextPart(TextFormat.Plain) With {.Text = Body.ToString}
                Wdbg("I", "Added body to FinalMessage.Body (plain text). Sending message...")
                SMTP_Client.Send(FinalMessage)
                Return True
            Catch ex As Exception
                Wdbg("E", "Failed to send message: {0}", ex.Message)
                WStkTrc(ex)
            End Try
            Return False
        End SyncLock
    End Function

    ''' <summary>
    ''' Sends a message with advanced features like attachments
    ''' </summary>
    ''' <param name="Recipient">Recipient name</param>
    ''' <param name="Subject">Subject</param>
    ''' <param name="Body">Body</param>
    ''' <returns>True if successful; False if unsuccessful.</returns>
    Public Function MailSendMessage(ByVal Recipient As String, ByVal Subject As String, ByVal Body As MimeEntity) As String
        SyncLock SMTP_Client.SyncRoot
            Try
                Dim FinalMessage As New MimeMessage
                FinalMessage.From.Add(MailboxAddress.Parse(Mail_Authentication.UserName))
                Wdbg("I", "Added sender to FinalMessage.From.")
                FinalMessage.To.Add(MailboxAddress.Parse(Recipient))
                Wdbg("I", "Added address to FinalMessage.To.")
                FinalMessage.Subject = Subject
                Wdbg("I", "Added subject to FinalMessage.Subject.")
                FinalMessage.Body = Body
                Wdbg("I", "Added body to FinalMessage.Body (plain text). Sending message...")
                SMTP_Client.Send(FinalMessage)
                Return True
            Catch ex As Exception
                Wdbg("E", "Failed to send message: {0}", ex.Message)
                WStkTrc(ex)
            End Try
            Return False
        End SyncLock
    End Function

    ''' <summary>
    ''' Sends an encrypted message with advanced features like attachments
    ''' </summary>
    ''' <param name="Recipient">Recipient name</param>
    ''' <param name="Subject">Subject</param>
    ''' <param name="Body">Body</param>
    ''' <returns>True if successful; False if unsuccessful.</returns>
    Public Function MailSendEncryptedMessage(ByVal Recipient As String, ByVal Subject As String, ByVal Body As MimeEntity) As String
        SyncLock SMTP_Client.SyncRoot
            Try
                Dim FinalMessage As New MimeMessage
                FinalMessage.From.Add(MailboxAddress.Parse(Mail_Authentication.UserName))
                Wdbg("I", "Added sender to FinalMessage.From.")
                FinalMessage.To.Add(MailboxAddress.Parse(Recipient))
                Wdbg("I", "Added address to FinalMessage.To.")
                FinalMessage.Subject = Subject
                Wdbg("I", "Added subject to FinalMessage.Subject.")
                FinalMessage.Body = MultipartEncrypted.Encrypt(New PGPContext, FinalMessage.To.Mailboxes, Body)
                Wdbg("I", "Added body to FinalMessage.Body (plain text). Sending message...")
                SMTP_Client.Send(FinalMessage)
                Return True
            Catch ex As Exception
                Wdbg("E", "Failed to send message: {0}", ex.Message)
                WStkTrc(ex)
            End Try
            Return False
        End SyncLock
    End Function

End Module

File: Kernel Simulator/Network/Mail/Transfer/MailTransfer.cs
﻿
// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Linq;
using System.Text;
using Extensification.StringExts;
using KS.ConsoleBase.Colors;
using KS.Kernel.Debugging;
using KS.Languages;
using KS.Misc.Writers.ConsoleWriters;
using KS.Network.Mail.Directory;
using KS.Network.Mail.PGP;
using KS.Shell.Shells.Mail;
using KS.TimeDate;
using MailKit;
using MailKit.Search;
using MimeKit;
using MimeKit.Cryptography;
using MimeKit.Text;

namespace KS.Network.Mail.Transfer
{
    public static class MailTransfer
    {

        /// <summary>
        /// Prints content of message to console
        /// </summary>
        /// <param name="MessageNum">Message number</param>
        /// <param name="Decrypt">Whether to decrypt messages or not</param>
        public static void MailPrintMessage(int MessageNum, bool Decrypt = false)
        {
            int Message = MessageNum - 1;
            int MaxMessagesIndex = MailShellCommon.IMAP_Messages.Count() - 1;
            DebugWriter.WriteDebug(DebugLevel.I, "Message number {0}", Message);
            if (Message < 0)
            {
                DebugWriter.WriteDebug(DebugLevel.E, "Trying to access message 0 or less than 0.");
                TextWriterColor.Write(Translate.DoTranslation("Message number may not be negative or zero."), true, ColorTools.ColTypes.Error);
                return;
            }
            else if (Message > MaxMessagesIndex)
            {
                DebugWriter.WriteDebug(DebugLevel.E, "Message {0} not in list. It was larger than MaxMessagesIndex ({1})", Message, MaxMessagesIndex);
                TextWriterColor.Write(Translate.DoTranslation("Message specified is not found."), true, ColorTools.ColTypes.Error);
                return;
            }

            lock (MailLogin.IMAP_Client.SyncRoot)
            {
                // Get message
                DebugWriter.WriteDebug(DebugLevel.I, "Getting message...");
                MimeMessage Msg;
                if (!string.IsNullOrEmpty(MailShellCommon.IMAP_CurrentDirectory) & !(MailShellCommon.IMAP_CurrentDirectory == "Inbox"))
                {
                    var Dir = MailDirectory.OpenFolder(MailShellCommon.IMAP_CurrentDirectory);
                    Msg = Dir.GetMessage(MailShellCommon.IMAP_Messages.ElementAtOrDefault(Message), default, MailShellCommon.Mail_Progress);
                }
                else
                {
                    Msg = MailLogin.IMAP_Client.Inbox.GetMessage(MailShellCommon.IMAP_Messages.ElementAtOrDefault(Message), default, MailShellCommon.Mail_Progress);
                }

                // Prepare view
                TextWriterColor.Write();

                // Print all the addresses that sent the mail
                DebugWriter.WriteDebug(DebugLevel.I, "{0} senders.", Msg.From.Count);
                foreach (InternetAddress Address in Msg.From)
                {
                    DebugWriter.WriteDebug(DebugLevel.I, "Address: {0} ({1})", Address.Name, Address.Encoding.EncodingName);
                    TextWriterColor.Write(Translate.DoTranslation("- From {0}"), true, ColorTools.ColTypes.ListEntry, Address.ToString());
                }

                // Print all the addresses that received the mail
                DebugWriter.WriteDebug(DebugLevel.I, "{0} receivers.", Msg.To.Count);
                foreach (InternetAddress Address in Msg.To)
                {
                    DebugWriter.WriteDebug(DebugLevel.I, "Address: {0} ({1})", Address.Name, Address.Encoding.EncodingName);
                    TextWriterColor.Write(Translate.DoTranslation("- To {0}"), true, ColorTools.ColTypes.ListEntry, Address.ToString());
                }

                // Print the date and time when the user received the mail
                DebugWriter.WriteDebug(DebugLevel.I, "Rendering time and date of {0}.", Msg.Date.DateTime.ToString());
                TextWriterColor.Write(Translate.DoTranslation("- Sent at {0} in {1}"), true, ColorTools.ColTypes.ListEntry, TimeDateRenderers.RenderTime(Msg.Date.DateTime), TimeDateRenderers.RenderDate(Msg.Date.DateTime));

                // Prepare subject
                TextWriterColor.Write();
                DebugWriter.WriteDebug(DebugLevel.I, "Subject length: {0}, {1}", Msg.Subject.Length, Msg.Subject);
                TextWriterColor.Write($"- {Msg.Subject}", false, ColorTools.ColTypes.ListEntry);

                // Write a sign after the subject if attachments are found
                DebugWriter.WriteDebug(DebugLevel.I, "Attachments count: {0}", Msg.Attachments.Count());
                if (Msg.Attachments.Count() > 0)
                {
                    TextWriterColor.Write(" - [*]", true, ColorTools.ColTypes.ListEntry);
                }
                else
                {
                    TextWriterColor.Write();
                }

                // Prepare body
                TextWriterColor.Write();
                DebugWriter.WriteDebug(DebugLevel.I, "Displaying body...");
                var DecryptedMessage = default(Dictionary<string, MimeEntity>);
                DebugWriter.WriteDebug(DebugLevel.I, "To decrypt: {0}", Decrypt);
                if (Decrypt)
                {
                    DecryptedMessage = DecryptMessage(Msg);
                    DebugWriter.WriteDebug(DebugLevel.I, "Decrypted messages length: {0}", DecryptedMessage.Count);
                    var DecryptedEntity = DecryptedMessage["Body"];
                    var DecryptedStream = new MemoryStream();
                    DebugWriter.WriteDebug(DebugLevel.I, $"Decrypted message type: {(DecryptedEntity is Multipart ? "Multipart" : "Singlepart")}");
                    if (DecryptedEntity is Multipart)
                    {
                        Multipart MultiEntity = (Multipart)DecryptedEntity;
                        DebugWriter.WriteDebug(DebugLevel.I, $"Decrypted message entity is {(MultiEntity is not null ? "multipart" : "nothing")}");
                        if (MultiEntity is not null)
                        {
                            for (int EntityNumber = 0, loopTo = MultiEntity.Count - 1; EntityNumber <= loopTo; EntityNumber++)
                            {
                                DebugWriter.WriteDebug(DebugLevel.I, $"Entity number {EntityNumber} is {(MultiEntity[EntityNumber].IsAttachment ? "an attachment" : "not an attachment")}");
                                if (!MultiEntity[EntityNumber].IsAttachment)
                                {
                                    MultiEntity[EntityNumber].WriteTo(DecryptedStream, true);
                                    DebugWriter.WriteDebug(DebugLevel.I, "Written {0} bytes to stream.", DecryptedStream.Length);
                                    DecryptedStream.Position = 0L;
                                    var DecryptedByte = new byte[(int)(DecryptedStream.Length + 1)];
                                    DecryptedStream.Read(DecryptedByte, 0, (int)DecryptedStream.Length);
                                    DebugWriter.WriteDebug(DebugLevel.I, "Written {0} bytes to buffer.", DecryptedByte.Length);
                                    TextWriterColor.Write(Encoding.Default.GetString(DecryptedByte), true, ColorTools.ColTypes.ListValue);
                                }
                            }
                        }
                    }
                    else
                    {
                        DecryptedEntity.WriteTo(DecryptedStream, true);
                        DebugWriter.WriteDebug(DebugLevel.I, "Written {0} bytes to stream.", DecryptedStream.Length);
                        DecryptedStream.Position = 0L;
                        var DecryptedByte = new byte[(int)(DecryptedStream.Length + 1)];
                        DecryptedStream.Read(DecryptedByte, 0, (int)DecryptedStream.Length);
                        DebugWriter.WriteDebug(DebugLevel.I, "Written {0} bytes to buffer.", DecryptedByte.Length);
                        TextWriterColor.Write(Encoding.Default.GetString(DecryptedByte), true, ColorTools.ColTypes.ListValue);
                    }
                }
                else
                {
                    TextWriterColor.Write(Msg.GetTextBody(MailShellCommon.Mail_TextFormat), true, ColorTools.ColTypes.ListValue);
                }
                TextWriterColor.Write();

                // Populate attachments
                if (Msg.Attachments.Count() > 0)
                {
                    TextWriterColor.Write(Translate.DoTranslation("Attachments:"), true, ColorTools.ColTypes.Neutral);
                    var AttachmentEntities = new List<MimeEntity>();
                    if (Decrypt)
                    {
                        DebugWriter.WriteDebug(DebugLevel.I, "Parsing attachments...");
                        for (int DecryptedEntityNumber = 0, loopTo1 = DecryptedMessage.Count - 1; DecryptedEntityNumber <= loopTo1; DecryptedEntityNumber++)
                        {
                            DebugWriter.WriteDebug(DebugLevel.I, "Is entity number {0} an attachment? {1}", DecryptedEntityNumber, DecryptedMessage.Keys.ElementAtOrDefault(DecryptedEntityNumber).Contains("Attachment"));
                            DebugWriter.WriteDebug(DebugLevel.I, "Is entity number {0} a body that is a multipart? {1}", DecryptedEntityNumber, DecryptedMessage.Keys.ElementAtOrDefault(DecryptedEntityNumber) == "Body" & DecryptedMessage["Body"] is Multipart);
                            if (DecryptedMessage.Keys.ElementAtOrDefault(DecryptedEntityNumber).Contains("Attachment"))
                            {
                                DebugWriter.WriteDebug(DebugLevel.I, "Adding entity {0} to attachment entities...", DecryptedEntityNumber);
                                AttachmentEntities.Add(DecryptedMessage.Values.ElementAtOrDefault(DecryptedEntityNumber));
                            }
                            else if (DecryptedMessage.Keys.ElementAtOrDefault(DecryptedEntityNumber) == "Body" & DecryptedMessage["Body"] is Multipart)
                            {
                                Multipart MultiEntity = (Multipart)DecryptedMessage["Body"];
                                DebugWriter.WriteDebug(DebugLevel.I, $"Decrypted message entity is {(MultiEntity is not null ? "multipart" : "nothing")}");
                                if (MultiEntity is not null)
                                {
                                    DebugWriter.WriteDebug(DebugLevel.I, "{0} entities found.", MultiEntity.Count);
                                    for (int EntityNumber = 0, loopTo2 = MultiEntity.Count - 1; EntityNumber <= loopTo2; EntityNumber++)
                                    {
                                        DebugWriter.WriteDebug(DebugLevel.I, $"Entity number {EntityNumber} is {(MultiEntity[EntityNumber].IsAttachment ? "an attachment" : "not an attachment")}");
                                        if (MultiEntity[EntityNumber].IsAttachment)
                                        {
                                            DebugWriter.WriteDebug(DebugLevel.I, "Adding entity {0} to attachment list...", EntityNumber);
                                            AttachmentEntities.Add(MultiEntity[EntityNumber]);
                                        }
                                    }
                                }
                            }
                        }
                    }
                    else
                    {
                        AttachmentEntities = (List<MimeEntity>)Msg.Attachments;
                    }
                    foreach (MimeEntity Attachment in AttachmentEntities)
                    {
                        DebugWriter.WriteDebug(DebugLevel.I, "Attachment ID: {0}", Attachment.ContentId);
                        if (Attachment is MessagePart)
                        {
                            DebugWriter.WriteDebug(DebugLevel.I, "Attachment is a message.");
                            TextWriterColor.Write($"- {Attachment.ContentDisposition?.FileName}", true, ColorTools.ColTypes.Neutral);
                        }
                        else
                        {
                            DebugWriter.WriteDebug(DebugLevel.I, "Attachment is a file.");
                            MimePart AttachmentPart = (MimePart)Attachment;
                            TextWriterColor.Write($"- {AttachmentPart.FileName}", true, ColorTools.ColTypes.Neutral);
                        }
                    }
                }
            }
        }

        /// <summary>
        /// Decrypts a message
        /// </summary>
        /// <param name="Text">Text part</param>
        /// <returns>A decrypted message, or null if unsuccessful.</returns>
        public static Dictionary<string, MimeEntity> DecryptMessage(MimeMessage Text)
        {
            var EncryptedDict = new Dictionary<string, MimeEntity>();
            DebugWriter.WriteDebug(DebugLevel.I, $"Encrypted message type: {(Text.Body is MultipartEncrypted ? "Multipart" : "Singlepart")}");
            if (Text.Body is MultipartEncrypted)
            {
                MultipartEncrypted Encrypted = (MultipartEncrypted)Text.Body;
                DebugWriter.WriteDebug(DebugLevel.I, $"Message type: {(Encrypted is not null ? "MultipartEncrypted" : "Nothing")}");
                DebugWriter.WriteDebug(DebugLevel.I, "Decrypting...");
                EncryptedDict.Add("Body", Encrypted.Decrypt(new PGPContext()));
            }
            else
            {
                DebugWriter.WriteDebug(DebugLevel.W, "Trying to decrypt plain text. Returning body...");
                EncryptedDict.Add("Body", Text.Body);
            }
            int AttachmentNumber = 1;
            foreach (MimeEntity TextAttachment in Text.Attachments)
            {
                DebugWriter.WriteDebug(DebugLevel.I, "Attachment number {0}", AttachmentNumber);
                DebugWriter.WriteDebug(DebugLevel.I, $"Encrypted attachment type: {(TextAttachment is MultipartEncrypted ? "Multipart" : "Singlepart")}");
                if (TextAttachment is MultipartEncrypted)
                {
                    MultipartEncrypted Encrypted = (MultipartEncrypted)TextAttachment;
                    DebugWriter.WriteDebug(DebugLevel.I, $"Attachment type: {(Encrypted is not null ? "MultipartEncrypted" : "Nothing")}");
                    DebugWriter.WriteDebug(DebugLevel.I, "Decrypting...");
                    EncryptedDict.Add("Attachment " + AttachmentNumber, Encrypted.Decrypt(new PGPContext()));
                }
                else
                {
                    DebugWriter.WriteDebug(DebugLevel.W, "Trying to decrypt plain attachment. Returning body...");
                    EncryptedDict.Add("Attachment " + AttachmentNumber, TextAttachment);
                }
                AttachmentNumber += 1;
            }
            return EncryptedDict;
        }

        /// <summary>
        /// Sends a message
        /// </summary>
        /// <param name="Recipient">Recipient name</param>
        /// <param name="Subject">Subject</param>
        /// <param name="Body">Body (only text. See <see cref="MailSendMessage(string, string, MimeEntity)"/> for more.)</param>
        /// <returns>True if successful; False if unsuccessful.</returns>
        public static string MailSendMessage(string Recipient, string Subject, string Body)
        {
            // Construct a message
            var FinalMessage = new MimeMessage();
            FinalMessage.From.Add(MailboxAddress.Parse(MailLogin.Mail_Authentication.UserName));
            DebugWriter.WriteDebug(DebugLevel.I, "Added sender to FinalMessage.From.");
            FinalMessage.To.Add(MailboxAddress.Parse(Recipient));
            DebugWriter.WriteDebug(DebugLevel.I, "Added address to FinalMessage.To.");
            FinalMessage.Subject = Subject;
            DebugWriter.WriteDebug(DebugLevel.I, "Added subject to FinalMessage.Subject.");
            FinalMessage.Body = new TextPart(TextFormat.Plain) { Text = Body.ToString() };
            DebugWriter.WriteDebug(DebugLevel.I, "Added body to FinalMessage.Body (plain text). Sending message...");

            // Send the message
            lock (MailLogin.SMTP_Client.SyncRoot)
            {
                try
                {
                    MailLogin.SMTP_Client.Send(FinalMessage, default, MailShellCommon.Mail_Progress);
                    return Convert.ToString(true);
                }
                catch (Exception ex)
                {
                    DebugWriter.WriteDebug(DebugLevel.E, "Failed to send message: {0}", ex.Message);
                    DebugWriter.WriteDebugStackTrace(ex);
                }
                return Convert.ToString(false);
            }
        }

        /// <summary>
        /// Sends a message with advanced features like attachments
        /// </summary>
        /// <param name="Recipient">Recipient name</param>
        /// <param name="Subject">Subject</param>
        /// <param name="Body">Body</param>
        /// <returns>True if successful; False if unsuccessful.</returns>
        public static string MailSendMessage(string Recipient, string Subject, MimeEntity Body)
        {
            // Construct a message
            var FinalMessage = new MimeMessage();
            FinalMessage.From.Add(MailboxAddress.Parse(MailLogin.Mail_Authentication.UserName));
            DebugWriter.WriteDebug(DebugLevel.I, "Added sender to FinalMessage.From.");
            FinalMessage.To.Add(MailboxAddress.Parse(Recipient));
            DebugWriter.WriteDebug(DebugLevel.I, "Added address to FinalMessage.To.");
            FinalMessage.Subject = Subject;
            DebugWriter.WriteDebug(DebugLevel.I, "Added subject to FinalMessage.Subject.");
            FinalMessage.Body = Body;
            DebugWriter.WriteDebug(DebugLevel.I, "Added body to FinalMessage.Body (plain text). Sending message...");

            // Send the message
            lock (MailLogin.SMTP_Client.SyncRoot)
            {
                try
                {
                    MailLogin.SMTP_Client.Send(FinalMessage, default, MailShellCommon.Mail_Progress);
                    return Convert.ToString(true);
                }
                catch (Exception ex)
                {
                    DebugWriter.WriteDebug(DebugLevel.E, "Failed to send message: {0}", ex.Message);
                    DebugWriter.WriteDebugStackTrace(ex);
                }
                return Convert.ToString(false);
            }
        }

        /// <summary>
        /// Sends an encrypted message with advanced features like attachments
        /// </summary>
        /// <param name="Recipient">Recipient name</param>
        /// <param name="Subject">Subject</param>
        /// <param name="Body">Body</param>
        /// <returns>True if successful; False if unsuccessful.</returns>
        public static string MailSendEncryptedMessage(string Recipient, string Subject, MimeEntity Body)
        {
            // Construct a message
            var FinalMessage = new MimeMessage();
            FinalMessage.From.Add(MailboxAddress.Parse(MailLogin.Mail_Authentication.UserName));
            DebugWriter.WriteDebug(DebugLevel.I, "Added sender to FinalMessage.From.");
            FinalMessage.To.Add(MailboxAddress.Parse(Recipient));
            DebugWriter.WriteDebug(DebugLevel.I, "Added address to FinalMessage.To.");
            FinalMessage.Subject = Subject;
            DebugWriter.WriteDebug(DebugLevel.I, "Added subject to FinalMessage.Subject.");
            FinalMessage.Body = MultipartEncrypted.Encrypt(new PGPContext(), FinalMessage.To.Mailboxes, Body);
            DebugWriter.WriteDebug(DebugLevel.I, "Added body to FinalMessage.Body (plain text). Sending message...");

            // Send the message
            lock (MailLogin.SMTP_Client.SyncRoot)
            {
                try
                {
                    MailLogin.SMTP_Client.Send(FinalMessage, default, MailShellCommon.Mail_Progress);
                    return Convert.ToString(true);
                }
                catch (Exception ex)
                {
                    DebugWriter.WriteDebug(DebugLevel.E, "Failed to send message: {0}", ex.Message);
                    DebugWriter.WriteDebugStackTrace(ex);
                }
                return Convert.ToString(false);
            }
        }

        /// <summary>
        /// Populates e-mail messages
        /// </summary>
        public static void PopulateMessages()
        {
            if (MailLogin.IMAP_Client.IsConnected)
            {
                lock (MailLogin.IMAP_Client.SyncRoot)
                {
                    if (string.IsNullOrEmpty(MailShellCommon.IMAP_CurrentDirectory) | MailShellCommon.IMAP_CurrentDirectory == "Inbox")
                    {
                        MailLogin.IMAP_Client.Inbox.Open(FolderAccess.ReadWrite);
                        DebugWriter.WriteDebug(DebugLevel.I, "Opened inbox");
                        MailShellCommon.IMAP_Messages = MailLogin.IMAP_Client.Inbox.Search(SearchQuery.All).Reverse();
                        DebugWriter.WriteDebug(DebugLevel.I, "Messages count: {0} messages", MailShellCommon.IMAP_Messages.LongCount());
                    }
                    else
                    {
                        var Folder = MailDirectory.OpenFolder(MailShellCommon.IMAP_CurrentDirectory);
                        DebugWriter.WriteDebug(DebugLevel.I, "Opened {0}", MailShellCommon.IMAP_CurrentDirectory);
                        MailShellCommon.IMAP_Messages = Folder.Search(SearchQuery.All).Reverse();
                        DebugWriter.WriteDebug(DebugLevel.I, "Messages count: {0} messages", MailShellCommon.IMAP_Messages.LongCount());
                    }
                }
            }
        }

    }
}
